### ISSUE TITLE
The quantized weights are different in differenent batches during evaluation

### ISSUE THREAD

**Issue by tangyuelm** [Xilinx/brevitas#153]
Hello, I want to figure out how this quantization approach works in real inference, so I first extracted the weights during evaluation using the codes below.
    class LeNet5(nn.Module):
        def __init__(self):
            super(LeNet5, self).__init__()
            self.features = self._make_layers()[0]
            self.classifier = self._make_layers()[1]
        def _make_layers(self):
            layersc = []
            layersf = []
            qconv1 = make_quant_conv2d(3, 6, kernel_size=5, stride=1, padding=0, groups=1, bias=False, bit_width=bitw)
            qconv1m = qconv1[0]
            qconv1w = qconv1[1]
            qconv1s = qconv1[2]
            layersc +=[qconv1m]
            qrelu1 = make_quant_relu(bit_width=bitw, return_quant_tensor=False)
            qrelu1m = qrelu1[0]
            qrelu1s = qrelu1[1]
            qpool1 = nn.MaxPool2d(kernel_size=2, stride=2)
            layersc += [qrelu1m,qpool1]
            qconv2 = make_quant_conv2d(6, 16, kernel_size=5, stride=1, padding=0, groups=1, bias=False, bit_width=bitw)
            qconv2m = qconv2[0]
            qconv2w = qconv2[1]
            qconv2s = qconv2[2]
            layersc += [qconv2m]
            qrelu2 = make_quant_relu(bit_width=bitw, return_quant_tensor=False)
            qrelu2m = qrelu2[0]
            qrelu2s = qrelu2[1]
            qpool2 = nn.MaxPool2d(kernel_size=2, stride=2)
            layersc += [qrelu2m, qpool2]
            qfc1 = make_quant_linear(400, 120, bias=False, bit_width=bitw, weight_scaling_per_output_channel=False)
            qfc1m = qfc1[0]
            qfc1w = qfc1[1]
            qfc1s = qfc1[2]
            layersf += [qfc1m]
            qrelu3 = make_quant_relu(bit_width=bitw, return_quant_tensor=True)
            qrelu3m = qrelu3[0]
            qrelu3s = qrelu3[1]
            layersf += [qrelu3m]
            qfc2 = make_quant_linear(120, 84, bias=False, bit_width=bitw, weight_scaling_per_output_channel=False)
            qfc2m = qfc2[0]
            qfc2w = qfc2[1]
            qfc2s = qfc2[2]
            layersf += [qfc2m]
            qrelu4 = make_quant_relu(bit_width=bitw, return_quant_tensor=True)
            qrelu4m = qrelu4[0]
            qrelu4s = qrelu4[1]
            layersf += [qrelu4m]
            qfc3 = make_quant_linear(84, 100, bias=False, bit_width=bitw, weight_scaling_per_output_channel=False)
            qfc3m = qfc3[0]
            qfc3w = qfc3[1]
            qfc3s = qfc3[2]
            layersf += [qfc3m]
            return [nn.Sequential(*layersc), nn.Sequential(*layersf),qfc3w,qfc3s]
        def save_weights(self):
            return self._make_layers(self)[2]

        def forward(self, x, printf):
            out = self.features(x)
            out = out.view(out.size(0), -1)
            out = self.classifier(out)
            if printf ==1:
                print("make_layers")
                print("fc3 weights:",self._make_layers()[2],"fc3 scale:",self._make_layers()[3])
            if printf ==2:
                print("make_layers")
                print("fc3 weights:", self._make_layers()[2], "fc3 scale:", self._make_layers()[3])
            return out

    def test(epoch,model):
        global best_acc
        model.eval()  
        test_loss = 0
        correct = 0
        total = 0
        top1 = AverageMeter()
        top5 = AverageMeter()
       
        with torch.no_grad():
            for batch_idx, (data, targets) in enumerate(test_dataloader1):
                data, targets = data.to(device), targets.to(device)
                if epoch <=1:
                    outputs = model(data,1)
                    print(model.state_dict()['module.classifier.4.weight'])
                    print(LeNet5.save_weights(LeNet5))
                if epoch > 1 and epoch < 199:
                    outputs = model(data,0)
                if epoch >= 199:
                    outputs = model(data,2)
                    print(model.state_dict()['module.classifier.4.weight'])
                    print(LeNet5.save_weights(LeNet5))
                loss = criterion(outputs, targets)
                test_loss += loss.item()
                _, predicted = outputs.max(1)
                total += targets.size(0)
                correct += predicted.eq(targets).sum().item()
               
                prec1, prec5 = accuracy(outputs.data, targets.data, topk=(1, 5))
                top1.update(prec1.item(), data.size(0))
                top5.update(prec5.item(), data.size(0))
            test_loss /= len(test_dataloader1.dataset)
            acc = 100. * correct / len(test_dataloader1.dataset)
            print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%),top1: {:.4f} , top5: {:.4f}\n'.format(
                test_loss, correct, len(test_dataloader1.dataset), acc, top1.avg, top5.avg))
        if acc > best_acc:
            best_acc = acc

However, I found the quantized weights are different in different batches during evaluation. The unquantized weights are the same. I think for real inference, the quantized weights should also remain the same, right? So could you please tell me why I have such a problem and how to solve it if I want to see the performance and accuracy in real inference? Thank you very much.

**Comment by tangyuelm** [comment]
I import the common.py file for the make quant function and display the weights and scale like:
def make_quant_conv2d(in_channels,
                      out_channels,
                      kernel_size,
                      stride,
                      padding,
                      groups,
                      bias,
                      bit_width,
                      enable_bias_quant=ENABLE_BIAS_QUANT,
                      weight_quant_type=QUANT_TYPE,
                      weight_scaling_impl_type=WEIGHT_SCALING_IMPL_TYPE,
                      weight_scaling_stats_op=WEIGHT_SCALING_STATS_OP,
                      weight_scaling_per_output_channel=WEIGHT_SCALING_PER_OUTPUT_CHANNEL,
                      weight_restrict_scaling_type=WEIGHT_RESTRICT_SCALING_TYPE,
                      weight_narrow_range=WEIGHT_NARROW_RANGE,
                      weight_scaling_min_val=SCALING_MIN_VAL):
    bias_quant_type = QUANT_TYPE if enable_bias_quant else QuantType.FP

    x= qnn.QuantConv2d(in_channels,
                           out_channels,
                           groups=groups,
                           kernel_size=kernel_size,
                           padding=padding,
                           stride=stride,
                           bias=bias,
                           bias_quant_type=bias_quant_type,
                           compute_output_bit_width=bias and enable_bias_quant,
                           compute_output_scale=bias and enable_bias_quant,
                           weight_bit_width=bit_width,
                           weight_quant_type=weight_quant_type,
                           weight_scaling_impl_type=weight_scaling_impl_type,
                           weight_scaling_stats_op=weight_scaling_stats_op,
                           weight_scaling_per_output_channel=weight_scaling_per_output_channel,
                           weight_restrict_scaling_type=weight_restrict_scaling_type,
                           weight_narrow_range=weight_narrow_range,
                           weight_scaling_min_val=weight_scaling_min_val)
    #x.save_weight()

    return [x,x.save_weights(),x.save_scale()]

**Comment by tangyuelm** [comment]
It seems that I make some mistakes to access the int weights. Now I am managing to solve it. But I am still not sure how to assign the weights and scales to a new model. Since I want to simulate the real inference, is it practical the pre-trained weights and scales to a brand new model?

**Comment by Giuseppe5** [comment]
Sorry for the late reply.
It is strange that quantized weights are different across batches but floating point weights are not, since quantized weights are computed starting from the FP version, but I see that you are managing to solve that issue. If you still have problems, leave us the code (as you already did, maybe try formatting it a little bit better) but also some basic script to run the model and test it.

Considering your second question, you can train your quantized model and afterward you can use the state_dict() method to access and save the weights of your model.
If then you restore the state_dict to another quantized model (with the same topology and structure, and quantization configuration), then everything will be equal to the original model.

Let me know if this clarifies your doubt, or there's something I missed!

**Comment by tangyuelm** [comment]
Thanks for your reply. The first problem is solved. I made some mistake accessing the int-weight. Now I am trying to assign the quantized weights to a new model since I want to simulate the real int8 inference but it has error now. Here is my current code:
    class LeNet5(nn.Module):
        def __init__(self):
            super(LeNet5, self).__init__()
            self.conv1 = make_quant_conv2d(3, 6, kernel_size=5, stride=1, padding=0, groups=1, bias=False, bit_width=bitw)[0]
            self.relu1 = make_quant_relu(bit_width=bitw, return_quant_tensor=False)[0]
            self.pool1 = nn.MaxPool2d(kernel_size=2, stride=2)
            self.conv2 = make_quant_conv2d(6, 16, kernel_size=5, stride=1, padding=0, groups=1, bias=False, bit_width=bitw)[0]  ##c3 conv:6 input feature maps kernel size 5*5,16 filters
            self.relu2 = make_quant_relu(bit_width=bitw, return_quant_tensor=False)[0]
            self.pool2 = nn.MaxPool2d(kernel_size=2, stride=2)
            self.fc1 = make_quant_linear(400, 120, bias=False, bit_width=bitw, weight_scaling_per_output_channel=False)[0]  ##c5 conv:16 IFMs, IFM size :1*1, kernel size 5*5, 120 filters
            self.relu3 = make_quant_relu(bit_width=bitw, return_quant_tensor=True)[0]
            self.fc2 = make_quant_linear(120, 84, bias=False, bit_width=bitw, weight_scaling_per_output_channel=False)[0]  ## F6 FC:120 IFMs, 84 OFMs
            self.relu4 = make_quant_relu(bit_width=bitw, return_quant_tensor=True)[0]
            self.fc3 = make_quant_linear(84, 100, bias=False, bit_width=bitw, weight_scaling_per_output_channel=False)[0]  ##Gaussian Connections 10 Outputs
            self.qconv1 = make_quant_conv2d(3, 6, kernel_size=5, stride=1, padding=0, groups=1, bias=False, bit_width=bitw)[0]
            self.qrelu1 = make_quant_relu(bit_width=bitw, return_quant_tensor=False)[0]
            self.qpool1 = nn.MaxPool2d(kernel_size=2, stride=2)
            self.qconv2 = make_quant_conv2d(6, 16, kernel_size=5, stride=1, padding=0, groups=1, bias=False, bit_width=bitw)[0]
            self.qrelu2 = make_quant_relu(bit_width=bitw, return_quant_tensor=False)[0]
            self.qpool2 = nn.MaxPool2d(kernel_size=2, stride=2)
            self.qfc1 = make_quant_linear(400, 120, bias=False, bit_width=bitw, weight_scaling_per_output_channel=False)[0]
            self.qrelu3 = make_quant_relu(bit_width=bitw, return_quant_tensor=True)[0]
            self.qfc2 = make_quant_linear(120, 84, bias=False, bit_width=bitw, weight_scaling_per_output_channel=False)[0]
            self.qrelu4 = make_quant_relu(bit_width=bitw, return_quant_tensor=True)[0]
            self.qfc3 = make_quant_linear(84, 100, bias=False, bit_width=bitw, weight_scaling_per_output_channel=False)[0]
       

        def forward(self, x,printf):
            outq =self.qconv1(x)
            outq = self.qpool1(self.qrelu1(outq))
            outq = self.qconv2(outq)
            outq = self.qpool2(self.qrelu2(outq))
            outq = outq.view(outq.size(0), -1)
            outq = self.qfc1(outq)
            outq = self.qrelu3(outq)
            outq = self.qfc2(outq)
            outq = self.qrelu4(outq)
            outq = self.qfc3(outq)
            if printf == 200:
                self.conv1.weight = torch.nn.Parameter(torch.Tensor(self.qconv1.int_weight))
                out1 = self.conv1(pack_quant_tensor(x,self.qconv1.quant_weight_scale,8))
                out1 = self.pool1(self.relu1(pack_quant_tensor(out1,self.qrelu1.quant_weight_scale, 8)))
                self.conv2.weight = torch.nn.Parameter(torch.Tensor(self.qconv2.int_weight))
                out1 = self.conv2(pack_quant_tensor(out1, self.qconv2.quant_weight_scale, 8))
                out1 = self.pool2(self.relu2(pack_quant_tensor(out1,self.qrelu2.quant_weight_scale, 8)))
                out1 = out1.view(out1.size(0), -1)
                self.fc1.weight = torch.nn.Parameter(torch.Tensor(self.qfc1.int_weight))
                out1 = self.fc1(pack_quant_tensor(out1, self.qfc1.quant_weight_scale, 8))
                out1 = self.relu3(pack_quant_tensor(out1,self.qrelu3.quant_weight_scale, 8))
                self.fc2.weight = torch.nn.Parameter(torch.Tensor(self.qfc2.int_weight))
                out1 = self.fc2(pack_quant_tensor(out1, self.qfc2.quant_weight_scale, 8))
                out1 = self.relu4(pack_quant_tensor(out1,self.qrelu4.quant_weight_scale, 8))
                self.fc3.weight = torch.nn.Parameter(torch.Tensor(self.qfc3.int_weight))
                out1 = self.fc3(pack_quant_tensor(out1, self.qfc3.quant_weight_scale, 8))
                print("fc3 weights:", self.fc3.int_weight, "fc3 scale:", self.fc3.quant_weight_scale)
                out= out1
            else:
                out = outq
            if printf ==1:
                print("make_layers")
                print("fc3 weights:",self.qfc3.int_weight,"fc3 scale:",self.qfc3.quant_weight_scale)
            if printf ==2:
                print("make_layers")
                print("fc3 weights:",self.qfc3.int_weight,"fc3 scale:",self.qfc3.quant_weight_scale)
            return out
I assign the weights from qconv to conv in the last training epoch but I have the error of "segmentation fault". Could you provide me any suggestions if I want to extract the int weights and assign them to a now model? Thank you.

**Comment by tangyuelm** [comment]
Hi, Giuseppe

I fixed the reported error, but the results are not right. 
In fact, what I want to do is to simulate the real inference after quantization because I am not sure how to implement the quantized inference on hardware like FPGA, and I am also wondering whether using Int instead of float point will influence the final accuracy for inference. 
Therefore, I plan to assign the quantized parameters to a new layer of the model (for example, after training, I can get the Int weights for a quantized conv layer. Then I can assign these weights to a common conv layer to simulate a convolution computation for inference.) However, I have no idea how the scales of the conv layer and activation layer work during the inference. The input of the dataset is float-point at the beginning right? Then after the first convolution layer, the output is also float-point even though the weights for the conv kernels are int, right? Then, how do the scale of the conv layer and the scale activation layer work to convert the output into Int in the real inference? 
Best Wishes,
Yue

**Comment by Giuseppe5** [comment]
Ok I understood what you're trying to do.
As of now, we are not supporting this feature, meaning that if you want to perform integer computation (on hardware, not simulated), you need to export your network according to the particular device you're using (you may want to look at [Finn](https://github.com/Xilinx/finn)).

We are working on a function similar to what you are saying, called Dynamic Quantization (look at #88 and #90), where at inference-time brevitas takes care of performing operations with integer or floating point according to the characteristics of the networks, inputs and so on, but it's not ready yet. 

In general, you need to consider that activations work with floating point values, so even if you have integer tensors, you should use the associated scale factor to re-cast them to floating point. Then to obtain your integer values, you simply divide by the activation's scale factor (performing a rounding if necessary). Related to the weight layers (such as convolution), you can try to extract the integer weights and assigning them to a "standard" pytorch layer, and then the output type will follow the input type (if input is float, output will be float, and the same with integers).
Having said that, you need to consider that you need to correctly propagate the scale factors, so that you can perform conversions between integers and floating point in the correct way, but this may be not trivial.

Our plan is to perform all these operations under the hood everytime a brevitas quantized network is used in eval() mode, so to have a "simulation" of what happens at inference time, and to speed-up the inference process (and avoid useless overhead of having to compute scale factors at inference time which are fixed).

I hope I was as exhaustive as possible, and I'm sorry I couldn't help you more. There may be the chance that I will resume working on that feature on the near future, so try to keep an eye on the threads I linked before.

**Comment by tangyuelm** [comment]
I see. Thanks for your explanation.
So you use linear quantization in brevitas, right? I did not find the formula about the relationship of the quantized Int value, the scaling factor, and the original parameters. For example, in a quantized Relu layer is the quantized output equals to  float  point activation/scale + activation(zero point), right? Then how to calculate the zero point? I see some parameters like SCALING_MIN_VAL, ACT_MAX_VAL, etc. but I don't know the meaning of this? Apart from the quant_act_scale, there is also a parameter called .act_quant_proxy.fused_activation_quant_proxy.tensor_quant.scaling_impl.learned_value? Are there any document about the meaning for the above-mentioned parameters?
Best Wishes,
Yue

**Comment by Giuseppe5** [comment]
We don't have yet any high-level documentation to explain how the quantization and the different parameters work in brevitas, but it is planned (as many other things).

[Here](https://github.com/Xilinx/brevitas/blob/49d5a1902c8f01521b344c0aae2d07ed672008ec/brevitas/core/quant.py#L681) you can find the actual formula for the quantization.

**Comment by tangyuelm** [comment]
Got it. Thank you very much.

**Comment by tangyuelm** [comment]
Hi, Giuseppe

I  followed your instruction to quantize and dequantize layer by layer during inference. The accuracy does not drop much. However, I can only quantize after the first convolution layer because I don't know how to quantize the input. I tried a few ways like calculating the scales and zero points for the inputs, using quant_hard_tanh functions, but all failed. I think quant_hard_tanh should be a choice because after I printing the activations of the convolutional computation after quant_hard_tanh the input, some values can match but some cannot. I am wondering whether it depends on the threshold. Now I use defaulted threshold which is 10.0.as assigned in your common.py. Can adjusting the threshold solve the problem? Are there any other parameters that influence the results?
Best Wishes,
Yue

**Comment by Giuseppe5** [comment]
You are mostly right, the best strategy for quantizing the input, if you have no other a-priori information, is to pass it through a QuantHardTanh. 
Of course, all the quantization hyper-parameters will influence this process and there may be some degradation (as in general with quantization). To limit this factor, you can play with different hyper-parameters. One is the threshold as you correctly said, but also the scaling factor implementation is very relevant (CONST, STATS, PARAMETER, the latter tends _in general_  to be better for activation) . 

For STATS and PARAMETER, the threshold is important only for the initialization of the scale factor, after that it will be updated according to the strategy selected, but of course a bad initialization may lead to non'convergence. In const, the value is kept fixed so it's much more relevant.

**Comment by tangyuelm** [comment]
I see. Thanks for your quick reply and detailed explanation.

**Comment by manwu1994** [comment]
hi, have you solve the above problem? I am also interested in implementing the quantized model on FPGA.
